File: basic_dsa_one_hundred_eight.java
/**
 * Basic DSA Problems - Two Sum II (Input Array Is Sorted)
 * Find two indices in sorted array that add up to target (1-indexed)
 */
import java.util.Arrays;

public class basic_dsa_one_hundred_eight {

    // Problem: Given 1-indexed sorted array and target, return indices [i, j] (i < j)
    // such that numbers[i] + numbers[j] == target. Exactly one solution exists.

    /**
     * Two pointers: left at start, right at end; if sum < target move left, else if sum > target move right.
     * Time O(n), Space O(1)
     */
    public static int[] twoSum(int[] numbers, int target) {
        if (numbers == null || numbers.length < 2) return new int[] { -1, -1 };
        int left = 0, right = numbers.length - 1;
        while (left < right) {
            int sum = numbers[left] + numbers[right];
            if (sum == target) return new int[] { left + 1, right + 1 };
            if (sum < target) left++;
            else right--;
        }
        return new int[] { -1, -1 };
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Two Sum II (Input Array Is Sorted)");
        System.out.println("=".repeat(60));

        System.out.println("[2,7,11,15], target=9 -> " + Arrays.toString(twoSum(new int[] { 2, 7, 11, 15 }, 9)));
        System.out.println("[2,3,4], target=6 -> " + Arrays.toString(twoSum(new int[] { 2, 3, 4 }, 6)));
        System.out.println("[-1,0], target=-1 -> " + Arrays.toString(twoSum(new int[] { -1, 0 }, -1)));
    }
}

File: basic_dsa_one_hundred_eight.py
"""
Basic DSA - Path Sum
Given the root of a binary tree and an integer targetSum, return True if the tree has a root-to-leaf path
such that adding up all the values along the path equals targetSum.
"""


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def has_path_sum(root, target_sum):
    """
    DFS accumulate remaining sum.
    Time Complexity: O(n)
    Space Complexity: O(h)
    """
    if not root:
        return False
    if not root.left and not root.right:
        return root.val == target_sum
    remaining = target_sum - root.val
    return has_path_sum(root.left, remaining) or has_path_sum(root.right, remaining)


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Path Sum")
    print("=" * 60)

    #       5
    #      / \
    #     4   8
    #    /   / \
    #   11  13  4
    #  /  \      \
    # 7    2      1
    root = TreeNode(
        5,
        TreeNode(4, TreeNode(11, TreeNode(7), TreeNode(2))),
        TreeNode(8, TreeNode(13), TreeNode(4, None, TreeNode(1))),
    )
    print(has_path_sum(root, 22))  # True
    print(has_path_sum(root, 26))  # True (5->8->13)
    print(has_path_sum(root, 18))  # True (5->8->4->1)

